GeosUtil/unitconv_mod.F90

```diff
@@ -2467,31 +2536,173 @@ SUBROUTINE ConvertBox_Kg_to_Kgm2( I, J, L, State_Chm, State_Grid, isAdjoint, RC
 !
 ! !LOCAL VARIABLES:
 !
-    INTEGER            :: N
+    ! Scalars
+    INTEGER            :: N,      S
+
+    ! Strings
     CHARACTER(LEN=255) :: errMsg, thisLoc
 
     !========================================================================
     ! ConvertBox_Kg_to_Kgm2 begins here!
     !========================================================================
 
-    ! Assume success
-    RC = GC_SUCCESS
+    ! Initialize
+    RC      = GC_SUCCESS
+    errMsg  = ''
+    thisLoc = &
+     ' -> at ConvertBox_Kg_to_Kgm2 (in GeosUtil/unitconv_mod.F90)'
+
+    ! Loop over species
+    !$OMP PARALLEL DO                                                        &
+    !$OMP DEFAULT( SHARED                                                   )&
+    !$OMP PRIVATE( S, N                                                     )
+    DO S = 1, SIZE( mapping )
+
+       ! Get the modelId from the mapping array
+       N = mapping(S)
+
+       ! Convert species concentration units
+       State_Chm%Species(N)%Conc(I,J,L) =                                    &
+       State_Chm%Species(N)%Conc(I,J,L) / State_Grid%Area_M2(I,J)
 
-    !$OMP PARALLEL DO       &
-    !$OMP DEFAULT( SHARED ) &
-    !$OMP PRIVATE( N      )
-    DO N = 1, State_Chm%nSpecies
-       State_Chm%Species(N)%Conc(I,J,L) =                             &
-                                  State_Chm%Species(N)%Conc(I,J,L)    &
-                                  / State_Grid%Area_M2(I,J)
 #ifdef ADJOINT
-       if (isAdjoint) &
-            State_Chm%SpeciesAdj(I,J,L,N) = State_Chm%SpeciesAdj(I,J,L,N) &
-                                  / State_Grid%Area_M2(I,J)
+       IF ( isAdjoint ) THEN
+          State_Chm%SpeciesAdj(I,J,L,N) =                                    &
+          State_Chm%SpeciesAdj(I,J,L,N) / State_Grid%Area_M2(I,J)
+       ENDIF
 #endif
+
+       ! Update units metadata
+       State_Chm%Species(N)%Previous_Units = State_Chm%Species(N)%Units
+       State_Chm%Species(N)%Units = KG_SPECIES_PER_M2
+
     ENDDO
     !$OMP END PARALLEL DO
 
   END SUBROUTINE ConvertBox_Kg_to_Kgm2
 !EOC
+!------------------------------------------------------------------------------
+!                  GEOS-Chem Global Chemical Transport Model                  !
+!------------------------------------------------------------------------------
+!BOP
+!
+! !IROUTINE: Check_Units
+!
+! !DESCRIPTION: Ensures that all species have the same Units value before
+!  unit conversion is done.
+!\\
+! !INTERFACE:
+!
+  SUBROUTINE Check_Units( State_Chm, mapping, units, RC )
```

Comment:
Would there be any benefit to making this a function returning true/false?

Reply:
@lizziel: This has been done in commit a80f173.  Having `Check_Units` be a function simplifies the code quite a bit.